### ISSUE TITLE
Pivot / unstack on large data frame does not work int32 overflow

### ISSUE THREAD

**Issue by MarkiesFredje** [pandas-dev/pandas#26314]
#### Code Sample, a copy-pastable example if possible

```python
predictor_purchases_p = predictor_purchases.groupby(["ag", "artikelnr"])["som"].max().unstack().fillna(0)

or

predictor_purchases_p = predictor_purchases.pivot(index="ag", columns="artikelnr", value="som")
```
#### Problem description
I'm working on rather large data (>100GB in memory) on a beefy server (3TB ram)
When refactoring my code from pandas 0.21 to latest version, the pivot / unstack now returns an exception.

Unstacked DataFrame is too big, causing int32 overflow

I was able to eliminate the problem by changing the reshape.py:
modify line 121from dtype np.int32 to dtype np.int64:
 `num_cells = np.multiply(num_rows, num_columns, dtype=np.int64)`

#### Expected Output

Not being limited by int32 dims when reshaping a data frame.
This feels like a restriction which should not be there.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.3.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-862.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.2
pytest: None
pip: 19.1
setuptools: 41.0.1
Cython: 0.29.7
numpy: 1.16.3
scipy: 1.2.1
pyarrow: None
xarray: 0.12.1
IPython: 7.5.0
sphinx: None
patsy: 0.5.1
dateutil: 2.8.0
pytz: 2019.1
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 3.0.3
openpyxl: 2.6.2
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: 1.3.3
pymysql: None
psycopg2: None
jinja2: 2.10.1
s3fs: None
fastparquet: 0.3.0
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

**Comment by jreback** [comment]
try on master this was recently patched

**Comment by MarkiesFredje** [comment]
Corporate server, so long story short, I can only work with conda releases.
I'm not able to pull anything off github and make a build.

**Comment by jreback** [comment]
https://github.com/pandas-dev/pandas/pull/23512/files

if you are trying to do this approach your problem differently

**Comment by jreback** [comment]
@Rblivingstone certainly would take a PR to allow int64 here; it’s not tested and that’s why we raise

**Comment by Code4SAFrankie** [comment]
I'm getting this error on 6Gb of memory use with pivoting the movielens large ratings.csv. So I agree in this day and age we need int64.

**Comment by Code4SAFrankie** [comment]
Turns out I get the same error even after changing the reshape.py line to `num_cells = np.multiply(num_rows, num_columns, dtype=np.int64)`, although the error definitely looks like it occurs there.

**Comment by buddingPop** [comment]
I get "ValueError: negative dimensions are not allowed" after changing the reshape.py line to num_cells = np.multiply(num_rows, num_columns, dtype=np.int64).

Any chance we have a different workaround? I only have 6000 columns...

**Comment by meganson** [comment]
I have same issue. I use 9.9 BG memory data. 

df.pivot_table(index='uno', columns=['programid'], values='avg_time')

ValueError: Unstacked DataFrame is too big, causing int32 overflow

**Comment by TomAugspurger** [comment]
@buddingPop @Code4SAFrankie @meganson are any of you interested in working on this?

We also need a reproducible example, if anyone can provide that.

**Comment by subhrajitb** [comment]
You can download the ratings.csv file from
https://www.kaggle.com/grouplens/movielens-20m-dataset

Then create a pivottable as below to reproduce the problem:
pivotTable = ratings.pivot_table(index=['userId'],columns=['movieId'],values='rating')

**Comment by TomAugspurger** [comment]
We won't be able to include downloading that dataset in a unit test. I assume this can be reproduced with the right style of random data? http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports

**Comment by meganson** [comment]
@TomAugspurger 
Here is my source.

ratings = pd.read_csv('/home/ml/rating_stg.csv').groupby(['uno', 'program_title'])['view_percent'].mean().unstack()
ratings.head()


---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-19a81104dd78> in <module>
----> 1 ratings = pd.read_csv('/home/ml/rating_stg.csv').groupby(['uno', 'program_title'])['view_percent'].mean().unstack()
      2 ratings.head()

~/anaconda3/lib/python3.7/site-packages/pandas/core/series.py in unstack(self, level, fill_value)
   3299         """
   3300         from pandas.core.reshape.reshape import unstack
-> 3301         return unstack(self, level, fill_value)
   3302 
   3303     # ----------------------------------------------------------------------

~/anaconda3/lib/python3.7/site-packages/pandas/core/reshape/reshape.py in unstack(obj, level, fill_value)
    394         unstacker = _Unstacker(obj.values, obj.index, level=level,
    395                                fill_value=fill_value,
--> 396                                constructor=obj._constructor_expanddim)
    397         return unstacker.get_result()
    398 

~/anaconda3/lib/python3.7/site-packages/pandas/core/reshape/reshape.py in __init__(self, values, index, level, value_columns, fill_value, constructor)
    122 
    123         if num_rows > 0 and num_columns > 0 and num_cells <= 0:
--> 124             raise ValueError('Unstacked DataFrame is too big, '
    125                              'causing int32 overflow')
    126 

ValueError: Unstacked DataFrame is too big, causing int32 overflow

**Comment by imanekho** [comment]
Hello,

i am facing the same issue trying to create a pivot table , if someone can help !

Thank you

**Comment by TomAugspurger** [comment]
We need a PR to fix it in pandas.

**Comment by imanekho** [comment]
this is the only solution you think? any alternative solution may be
i am working on my final project and i have a large dataset to late to change it 
thank yiu

**Comment by ericbrown** [comment]
> this is the only solution you think? any alternative solution may be
> i am working on my final project and i have a large dataset to late to change it
> thank yiu

You could attempt to downgrade your pandas to 0.21 (See here https://github.com/pandas-dev/pandas/issues/26314#issuecomment-490502770).  I ran into a similar problem recently and downgraded and it worked on one of my datasets.  It isn't the best solution but it might get you moving.

**Comment by imanekho** [comment]
thank you

**Comment by imanekho** [comment]
Can i have an idea how you dwnograded your version i tried to intsall the 21.0 that i get alot of error on denpendencies with other libraries 
Thank you

**Comment by ericbrown** [comment]
> Can i have an idea how you dwnograded your version i tried to intsall the 21.0 that i get alot of error on denpendencies with other libraries
> Thank you

 pip install pandas==0.21

**Comment by imanekho** [comment]
this is exactly the command i used

**Comment by karthikktn** [comment]
I got 48000 columns, I am getting the same error
ValueError: Unstacked DataFrame is too big, causing int32 overflow

Is there is any limit to columns in 0.24.2

**Comment by atriantafybbc** [comment]
> this is exactly the command i used

try with Python 3.6

**Comment by dumbledad** [comment]
Any news on this, I'm failing to roll back to 0.21. My next course of action is to rewrite vectorised function as a loop but I'd rather not.

**Comment by jimhavrilla** [comment]
I got this error in ver. `0.25.3`, any news on it?

**Comment by TomAugspurger** [comment]
Still open, still looking for a volunteer to fix it.

**Comment by pavany666666** [comment]
Any update on this? 
Does downgrading the Pandas help??

**Comment by jreback** [comment]
pandas is an all volunteer project; PRs would be accepted from the community for this or any other issues

**Comment by pavany666666** [comment]
@KaonToPion Please let us know once you checkin... I would like to have this fix ASAP... Thanks a lot!

**Comment by KaonToPion** [comment]
I am sorry, I am having issues with installing the pandas development enviroment that's why it's taking long

**Comment by TomAugspurger** [comment]
https://pandas.pydata.org/docs/development/contributing.html, in case it helps.

**Comment by storopoli** [comment]
@KaonToPion also waiting on this fix. If you could please take another effort a try that would be great. Thank you!

**Comment by ghuls** [comment]
The current code to check for int32 overflow is wrong anyway as it will not catch all overflows (only when the result is still negative):
https://github.com/pandas-dev/pandas/blob/a9cacd9efee31e6cff28125ffbad00fad1617833/pandas/core/reshape/reshape.py#L115-L119

Correct check:
```python
# GH20601: Catch int32 overflow if the number of cells is too high. 
if num_rows > 0 and num_columns > 0 and num_rows * num_columns > 2**31 - 1: 
    raise ValueError("Unstacked DataFrame is too big, causing int32 overflow") 
```

Examples:
```python
In [1]: import numpy as np

In [2]: def test_overflow(num_rows, num_columns):
     ...:     num_cells = np.multiply(num_rows, num_columns, dtype=np.int32)
     ...:     if num_rows > 0 and num_columns > 0 and num_cells <= 0:
     ...:         print("Unstacked DataFrame is too big, causing int32 overflow (np.multiply)")
     ...:     if num_rows > 0 and num_columns > 0 and num_rows * num_columns > 2**31 - 1:
     ...:         print("Unstacked DataFrame is too big, causing int32 overflow (python)")
     ...: 
In [3]: test_overflow(400_000, 1_000)

In [4]: test_overflow(4_000_000, 1_000)
Unstacked DataFrame is too big, causing int32 overflow (np.multiply)
Unstacked DataFrame is too big, causing int32 overflow (python)

In [5]: test_overflow(40_000_000, 1_000)
Unstacked DataFrame is too big, causing int32 overflow (python)

In [6]: test_overflow(400_000_000, 1_000)
Unstacked DataFrame is too big, causing int32 overflow (python)
```

**Comment by kamal1262** [comment]
I faced the same issue and needed to have an urgent fix for the unexpected int32 overflow.  One of the recommendation engine using lightFM was running in production and at some point number of users base was more than 7 million records with around 21k items. 

So, I chunked the dataset, create the pivot table using **unstack** and append to a global dataframe. 

`chunk_size = 50000
chunks = [x for x in range(0, df.shape[0], chunk_size)]

for i in range(0, len(chunks) - 1):
    print(chunks[i], chunks[i + 1] - 1)
0 49999
50000 99999
100000 149999
150000 199999
200000 249990
.........................
`

`pivot_df = pd.DataFrame()

for i in tqdm(range(0, len(chunks) - 1)):
    chunk_df = df_t.iloc[ chunks[i]:chunks[i + 1] - 1]
    interactions = (
    chunk_df.groupby([user_col, item_col])[rating_col]
    .sum()
    .unstack()
    .reset_index()
    .fillna(0)
    .set_index(user_col)
    )
    print (interactions.shape)
    pivot_df = pivot_df.append(interactions, sort=False)`


And then I have to make a sparse matrix as input to lightFM recommendation model (run matrix-factorization algorithm). You can use it for any use case where unstacking is required. Using the following code, converted to sparse matrix- 

`from scipy import sparse
sparse_matrix = sparse.csr_matrix(df_new.to_numpy())`

NB: Pandas has _pivot_table_ function which can be used for unstacking if your data is small. For my case, _pivot_table_ was really slow.

**Comment by lucashadin** [comment]
Did this ever get patched? :)

**Comment by pavany666666** [comment]
nope. I don't think so.I moved to alternate ways of doing it.
    On Monday, 14 June, 2021, 04:08:57 pm GMT-4, lucashadin ***@***.***> wrote:  
 
 


Did this ever get patched? :)

**Comment by lucashadin** [comment]
> nope. I don't think so.I moved to alternate ways of doing it. On Monday, 14 June, 2021, 04:08:57 pm GMT-4, lucashadin ***@***.***> wrote: Did this ever get patched? :) — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or unsubscribe.

How did you solve it? Did you use the chunking method that kamal1262 mentioned? Or another package? Any advice would be great :)
